reward/staking_manager.go

```diff
@@ -40,65 +41,89 @@ type blockChain interface {
-		logger.Debug("StakingInfoCache hit.", "Block number", blockNum, "number of staking block", stakingBlockNumber)
+	// Get staking info from cache
+	if cachedStakingInfo := sm.stakingInfoCache.get(stakingBlockNumber); cachedStakingInfo != nil {
+		logger.Debug("StakingInfoCache hit.", "Block number", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", cachedStakingInfo)
 		return cachedStakingInfo
 	}
 
-	stakingInfo, err := sm.updateStakingCache(stakingBlockNumber)
+	// Get staking info from db
+	if sm.dbManager != nil {
+		if storedStakingInfo, err := sm.dbManager.ReadStakingInfo(stakingBlockNumber); storedStakingInfo != nil && err == nil {
+			logger.Debug("StakingInfoDB hit.", "Block number", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", storedStakingInfo)
+			return storedStakingInfo.(*StakingInfo)
+		} else {
+			logger.Warn("Failed to get staking info from db.", "blockNum", blockNum, "err", err, "value", storedStakingInfo)
+		}
+	}
+
+	// Get staking info from block header and updates it to cache and db
+	calcStakingInfo, err := sm.updateStakingInfo(stakingBlockNumber)
 	if err != nil {
-		logger.Error("Failed to get stakingInfo", "Block number", blockNum, "number of staking block", stakingBlockNumber, "err", err)
+		logger.Error("Failed to get stakingInfo", "Block number", blockNum, "staking block number", stakingBlockNumber, "err", err, "stakingInfo", calcStakingInfo)
 		return nil
 	}
 
-	logger.Debug("Complete StakingInfoCache update.", "Block number", blockNum, "number of staking block", stakingBlockNumber)
-	return stakingInfo
+	logger.Debug("Get stakingInfo from header.", "Block number", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", calcStakingInfo)
+	return calcStakingInfo
 }
 
 func (sm *StakingManager) IsActivated() bool {
 	return sm.isActivated
 }
 
-// updateStakingCache updates staking cache with a stakingInfo of a given block number.
-func (sm *StakingManager) updateStakingCache(blockNum uint64) (*StakingInfo, error) {
-	stakingInfo, err := sm.ac.getStakingInfoFromAddressBook(blockNum)
+// updateStakingInfo updates staking info in cache and db created from given block number.
+func (sm *StakingManager) updateStakingInfo(blockNum uint64) (*StakingInfo, error) {
+	stakingInfo, err := sm.addressBookConnector.getStakingInfoFromAddressBook(blockNum)
 	if err != nil {
 		return nil, err
 	}
 	sm.isActivated = true
-	sm.sic.add(stakingInfo)
 
-	logger.Info("Add a new stakingInfo to the stakingInfoCache", "stakingInfo", stakingInfo)
+	// update cache
+	sm.stakingInfoCache.add(stakingInfo)
+
+	// update db
+	if sm.dbManager != nil {
+		err := sm.dbManager.WriteStakingInfo(stakingInfo)
+		if err != nil {
+			logger.Warn("Failed to write staking info to db.", "err", err, "stakingInfo", stakingInfo)
+		}
+	}
+
+	logger.Info("Add a new stakingInfo to stakingInfoCache and stakingInfoDB", "stakingInfo", stakingInfo)
```

Comment:
I think `stakingInfo` is raw data. So I think `Debug` is more proper. What do you think?
```suggestion
	logger.Info("Add a new stakingInfo to stakingInfoCache and stakingInfoDB", "blockNum", blockNum)
	logger.Debug("Added stakingInfo", "stakingInfo", stakingInfo)
```

Reply:
Review reflected